File: services/caption_service.py
"""
Caption service for caption editing operations.

Handles:
- Add trigger words to captions
- Remove specific tags
- Replace text (with regex support)
- Read/write individual captions
- Bulk operations on entire datasets
"""

import re
import logging
from pathlib import Path
from typing import List, Optional

from services.models.caption import (
    AddTriggerWordRequest,
    RemoveTagsRequest,
    ReplaceTextRequest,
    ReadCaptionRequest,
    WriteCaptionRequest,
    CaptionOperationResponse,
    CaptionReadResponse,
)
from services.core.exceptions import ValidationError, NotFoundError
from services.core.validation import validate_dataset_path, ALLOWED_IMAGE_EXTENSIONS

logger = logging.getLogger(__name__)


class CaptionService:
    """
    High-level service for caption editing.

    Responsibilities:
    - Bulk caption operations
    - Individual caption read/write
    - Tag manipulation
    - Text replacement
    """

    async def add_trigger_word(self, request: AddTriggerWordRequest) -> CaptionOperationResponse:
        """
        Add a trigger word to all captions in a dataset.

        Args:
            request: Trigger word addition request

        Returns:
            CaptionOperationResponse with count of modified files
        """
        try:
            dataset_path = validate_dataset_path(request.dataset_dir)

            if not dataset_path.exists():
                raise NotFoundError(f"Dataset not found: {request.dataset_dir}")

            files_modified = 0
            errors = []

            # Process all caption files
            for image_file in dataset_path.iterdir():
                if not image_file.is_file():
                    continue
                if image_file.suffix.lower() not in ALLOWED_IMAGE_EXTENSIONS:
                    continue

                caption_file = image_file.with_suffix(request.caption_extension)

                try:
                    # Read existing caption or create empty
                    if caption_file.exists():
                        caption_text = caption_file.read_text(encoding='utf-8').strip()
                    else:
                        caption_text = ""

                    # Add trigger word
                    if request.position == "start":
                        new_caption = f"{request.trigger_word}, {caption_text}" if caption_text else request.trigger_word
                    else:  # end
                        new_caption = f"{caption_text}, {request.trigger_word}" if caption_text else request.trigger_word

                    # Write back
                    caption_file.write_text(new_caption, encoding='utf-8')
                    files_modified += 1

                except Exception as e:
                    errors.append(f"{image_file.name}: {str(e)}")

            logger.info(f"Added trigger word '{request.trigger_word}' to {files_modified} captions")

            return CaptionOperationResponse(
                success=True,
                message=f"Added trigger word to {files_modified} captions",
                files_modified=files_modified,
                errors=errors
            )

        except (ValidationError, NotFoundError) as e:
            return CaptionOperationResponse(
                success=False,
                message=str(e),
                files_modified=0,
                errors=[str(e)]
            )

    async def remove_tags(self, request: RemoveTagsRequest) -> CaptionOperationResponse:
        """
        Remove specific tags from all captions in a dataset.

        Args:
            request: Tag removal request

        Returns:
            CaptionOperationResponse with count of modified files
        """
        try:
            dataset_path = validate_dataset_path(request.dataset_dir)

            if not dataset_path.exists():
                raise NotFoundError(f"Dataset not found: {request.dataset_dir}")

            files_modified = 0
            errors = []

            # Process all caption files
            for caption_file in dataset_path.glob(f"*{request.caption_extension}"):
                try:
                    caption_text = caption_file.read_text(encoding='utf-8').strip()

                    # Split into tags
                    tags = [tag.strip() for tag in caption_text.split(',')]

                    # Remove unwanted tags (case-insensitive)
                    tags_lower = {tag.lower() for tag in request.tags_to_remove}
                    filtered_tags = [tag for tag in tags if tag.lower() not in tags_lower]

                    # Write back if changed
                    if len(filtered_tags) != len(tags):
                        new_caption = ', '.join(filtered_tags)
                        caption_file.write_text(new_caption, encoding='utf-8')
                        files_modified += 1

                except Exception as e:
                    errors.append(f"{caption_file.name}: {str(e)}")

            logger.info(f"Removed tags from {files_modified} captions")

            return CaptionOperationResponse(
                success=True,
                message=f"Removed tags from {files_modified} captions",
                files_modified=files_modified,
                errors=errors
            )

        except (ValidationError, NotFoundError) as e:
            return CaptionOperationResponse(
                success=False,
                message=str(e),
                files_modified=0,
                errors=[str(e)]
            )

    async def replace_text(self, request: ReplaceTextRequest) -> CaptionOperationResponse:
        """
        Replace text in all captions (with optional regex support).

        Args:
            request: Text replacement request

        Returns:
            CaptionOperationResponse with count of modified files
        """
        try:
            dataset_path = validate_dataset_path(request.dataset_dir)

            if not dataset_path.exists():
                raise NotFoundError(f"Dataset not found: {request.dataset_dir}")

            files_modified = 0
            errors = []

            # Process all caption files
            for caption_file in dataset_path.glob(f"*{request.caption_extension}"):
                try:
                    caption_text = caption_file.read_text(encoding='utf-8')

                    # Replace text
                    if request.use_regex:
                        new_caption = re.sub(
                            request.find_text,
                            request.replace_text,
                            caption_text
                        )
                    else:
                        new_caption = caption_text.replace(
                            request.find_text,
                            request.replace_text
                        )

                    # Write back if changed
                    if new_caption != caption_text:
                        caption_file.write_text(new_caption, encoding='utf-8')
                        files_modified += 1

                except Exception as e:
                    errors.append(f"{caption_file.name}: {str(e)}")

            logger.info(f"Replaced text in {files_modified} captions")

            return CaptionOperationResponse(
                success=True,
                message=f"Replaced text in {files_modified} captions",
                files_modified=files_modified,
                errors=errors
            )

        except (ValidationError, NotFoundError) as e:
            return CaptionOperationResponse(
                success=False,
                message=str(e),
                files_modified=0,
                errors=[str(e)]
            )

    async def read_caption(self, request: ReadCaptionRequest) -> CaptionReadResponse:
        """
        Read a single caption file.

        Args:
            request: Caption read request

        Returns:
            CaptionReadResponse with caption content
        """
        try:
            image_path = Path(request.image_path)
            caption_path = image_path.with_suffix(request.caption_extension)

            if caption_path.exists():
                caption_text = caption_path.read_text(encoding='utf-8')
                return CaptionReadResponse(
                    success=True,
                    image_path=str(image_path),
                    caption_path=str(caption_path),
                    caption_text=caption_text,
                    exists=True
                )
            else:
                return CaptionReadResponse(
                    success=True,
                    image_path=str(image_path),
                    caption_path=str(caption_path),
                    caption_text=None,
                    exists=False
                )

        except Exception as e:
            logger.error(f"Failed to read caption: {e}")
            return CaptionReadResponse(
                success=False,
                image_path=request.image_path,
                caption_path=str(Path(request.image_path).with_suffix(request.caption_extension)),
                caption_text=None,
                exists=False
            )

    async def write_caption(self, request: WriteCaptionRequest) -> CaptionOperationResponse:
        """
        Write a single caption file.

        Args:
            request: Caption write request

        Returns:
            CaptionOperationResponse
        """
        try:
            image_path = Path(request.image_path)
            caption_path = image_path.with_suffix(request.caption_extension)

            # Ensure parent directory exists
            caption_path.parent.mkdir(parents=True, exist_ok=True)

            # Write caption
            caption_path.write_text(request.caption_text, encoding='utf-8')

            logger.info(f"Wrote caption: {caption_path.name}")

            return CaptionOperationResponse(
                success=True,
                message=f"Caption written to {caption_path.name}",
                files_modified=1,
                errors=[]
            )

        except Exception as e:
            logger.error(f"Failed to write caption: {e}")
            return CaptionOperationResponse(
                success=False,
                message=f"Failed to write caption: {e}",
                files_modified=0,
                errors=[str(e)]
            )


# Global service instance
caption_service = CaptionService()

File: frontend/lib/node-services/caption-service.ts
/**
 * Caption Service - Node.js caption editing operations
 * Migrated from Python services/caption_service.py
 *
 * Handles:
 * - Add trigger words to captions
 * - Remove specific tags
 * - Replace text (with regex support)
 * - Read/write individual captions
 * - Bulk operations on entire datasets
 */

import fs from 'fs/promises';
import path from 'path';

// ========== Types ==========

export interface AddTriggerWordRequest {
  dataset_dir: string;
  trigger_word: string;
  position: 'start' | 'end';
  caption_extension?: string;
}

export interface RemoveTagsRequest {
  dataset_dir: string;
  tags_to_remove: string[];
  caption_extension?: string;
}

export interface ReplaceTextRequest {
  dataset_dir: string;
  find_text: string;
  replace_text: string;
  use_regex?: boolean;
  caption_extension?: string;
}

export interface ReadCaptionRequest {
  image_path: string;
  caption_extension?: string;
}

export interface WriteCaptionRequest {
  image_path: string;
  caption_text: string;
  caption_extension?: string;
}

export interface CaptionOperationResponse {
  success: boolean;
  message: string;
  files_modified: number;
  errors: string[];
}

export interface CaptionReadResponse {
  success: boolean;
  image_path: string;
  caption_path: string;
  caption_text: string | null;
  exists: boolean;
}

// ========== Configuration ==========

const ALLOWED_IMAGE_EXTENSIONS = ['.jpg', '.jpeg', '.png', '.gif', '.webp', '.bmp'];
const DEFAULT_CAPTION_EXTENSION = '.txt';

// ========== Helper Functions ==========

/**
 * Validate dataset path exists and is a directory
 */
async function validateDatasetPath(datasetPath: string): Promise<string> {
  try {
    const stats = await fs.stat(datasetPath);
    if (!stats.isDirectory()) {
      throw new Error('Path is not a directory');
    }
    return path.resolve(datasetPath);
  } catch (error) {
    if (error instanceof Error && error.message.includes('ENOENT')) {
      throw new Error(`Dataset not found: ${datasetPath}`);
    }
    throw error;
  }
}

/**
 * Check if file is an image based on extension
 */
function isImageFile(filePath: string): boolean {
  const ext = path.extname(filePath).toLowerCase();
  return ALLOWED_IMAGE_EXTENSIONS.includes(ext);
}

/**
 * Get all image files in a directory
 */
async function getImageFiles(dirPath: string): Promise<string[]> {
  const entries = await fs.readdir(dirPath, { withFileTypes: true });
  return entries
    .filter(entry => entry.isFile() && isImageFile(entry.name))
    .map(entry => path.join(dirPath, entry.name));
}

/**
 * Get all caption files in a directory
 */
async function getCaptionFiles(dirPath: string, extension: string): Promise<string[]> {
  const entries = await fs.readdir(dirPath, { withFileTypes: true });
  return entries
    .filter(entry => entry.isFile() && entry.name.endsWith(extension))
    .map(entry => path.join(dirPath, entry.name));
}

// ========== CaptionService Class ==========

export class CaptionService {
  /**
   * Add a trigger word to all captions in a dataset
   */
  async addTriggerWord(request: AddTriggerWordRequest): Promise<CaptionOperationResponse> {
    try {
      const datasetPath = await validateDatasetPath(request.dataset_dir);
      const captionExt = request.caption_extension || DEFAULT_CAPTION_EXTENSION;

      let filesModified = 0;
      const errors: string[] = [];

      // Get all image files
      const imageFiles = await getImageFiles(datasetPath);

      for (const imageFile of imageFiles) {
        try {
          // Get corresponding caption file
          const captionFile = imageFile.replace(path.extname(imageFile), captionExt);

          // Read existing caption or create empty
          let captionText = '';
          try {
            captionText = (await fs.readFile(captionFile, 'utf-8')).trim();
          } catch {
            // Caption file doesn't exist, start with empty string
          }

          // Add trigger word
          let newCaption: string;
          if (request.position === 'start') {
            newCaption = captionText
              ? `${request.trigger_word}, ${captionText}`
              : request.trigger_word;
          } else {
            // end
            newCaption = captionText
              ? `${captionText}, ${request.trigger_word}`
              : request.trigger_word;
          }

          // Write back
          await fs.writeFile(captionFile, newCaption, 'utf-8');
          filesModified++;
        } catch (error) {
          errors.push(
            `${path.basename(imageFile)}: ${error instanceof Error ? error.message : String(error)}`
          );
        }
      }

      return {
        success: true,
        message: `Added trigger word to ${filesModified} captions`,
        files_modified: filesModified,
        errors,
      };
    } catch (error) {
      return {
        success: false,
        message: error instanceof Error ? error.message : String(error),
        files_modified: 0,
        errors: [error instanceof Error ? error.message : String(error)],
      };
    }
  }

  /**
   * Remove specific tags from all captions in a dataset
   */
  async removeTags(request: RemoveTagsRequest): Promise<CaptionOperationResponse> {
    try {
      const datasetPath = await validateDatasetPath(request.dataset_dir);
      const captionExt = request.caption_extension || DEFAULT_CAPTION_EXTENSION;

      let filesModified = 0;
      const errors: string[] = [];

      // Get all caption files
      const captionFiles = await getCaptionFiles(datasetPath, captionExt);

      // Create lowercase set for case-insensitive matching
      const tagsToRemoveLower = new Set(request.tags_to_remove.map(tag => tag.toLowerCase()));

      for (const captionFile of captionFiles) {
        try {
          const captionText = (await fs.readFile(captionFile, 'utf-8')).trim();

          // Split into tags
          const tags = captionText.split(',').map(tag => tag.trim());

          // Remove unwanted tags (case-insensitive)
          const filteredTags = tags.filter(tag => !tagsToRemoveLower.has(tag.toLowerCase()));

          // Write back if changed
          if (filteredTags.length !== tags.length) {
            const newCaption = filteredTags.join(', ');
            await fs.writeFile(captionFile, newCaption, 'utf-8');
            filesModified++;
          }
        } catch (error) {
          errors.push(
            `${path.basename(captionFile)}: ${error instanceof Error ? error.message : String(error)}`
          );
        }
      }

      return {
        success: true,
        message: `Removed tags from ${filesModified} captions`,
        files_modified: filesModified,
        errors,
      };
    } catch (error) {
      return {
        success: false,
        message: error instanceof Error ? error.message : String(error),
        files_modified: 0,
        errors: [error instanceof Error ? error.message : String(error)],
      };
    }
  }

  /**
   * Replace text in all captions (with optional regex support)
   */
  async replaceText(request: ReplaceTextRequest): Promise<CaptionOperationResponse> {
    try {
      const datasetPath = await validateDatasetPath(request.dataset_dir);
      const captionExt = request.caption_extension || DEFAULT_CAPTION_EXTENSION;

      let filesModified = 0;
      const errors: string[] = [];

      // Get all caption files
      const captionFiles = await getCaptionFiles(datasetPath, captionExt);

      for (const captionFile of captionFiles) {
        try {
          const captionText = await fs.readFile(captionFile, 'utf-8');

          // Replace text
          let newCaption: string;
          if (request.use_regex) {
            const regex = new RegExp(request.find_text, 'g');
            newCaption = captionText.replace(regex, request.replace_text);
          } else {
            newCaption = captionText.split(request.find_text).join(request.replace_text);
          }

          // Write back if changed
          if (newCaption !== captionText) {
            await fs.writeFile(captionFile, newCaption, 'utf-8');
            filesModified++;
          }
        } catch (error) {
          errors.push(
            `${path.basename(captionFile)}: ${error instanceof Error ? error.message : String(error)}`
          );
        }
      }

      return {
        success: true,
        message: `Replaced text in ${filesModified} captions`,
        files_modified: filesModified,
        errors,
      };
    } catch (error) {
      return {
        success: false,
        message: error instanceof Error ? error.message : String(error),
        files_modified: 0,
        errors: [error instanceof Error ? error.message : String(error)],
      };
    }
  }

  /**
   * Read a single caption file
   */
  async readCaption(request: ReadCaptionRequest): Promise<CaptionReadResponse> {
    try {
      const imagePath = path.resolve(request.image_path);
      const captionExt = request.caption_extension || DEFAULT_CAPTION_EXTENSION;
      const captionPath = imagePath.replace(path.extname(imagePath), captionExt);

      try {
        const captionText = await fs.readFile(captionPath, 'utf-8');
        return {
          success: true,
          image_path: imagePath,
          caption_path: captionPath,
          caption_text: captionText,
          exists: true,
        };
      } catch {
        // Caption file doesn't exist
        return {
          success: true,
          image_path: imagePath,
          caption_path: captionPath,
          caption_text: null,
          exists: false,
        };
      }
    } catch (error) {
      return {
        success: false,
        image_path: request.image_path,
        caption_path: path.resolve(request.image_path).replace(path.extname(request.image_path), request.caption_extension || DEFAULT_CAPTION_EXTENSION),
        caption_text: null,
        exists: false,
      };
    }
  }

  /**
   * Write a single caption file
   */
  async writeCaption(request: WriteCaptionRequest): Promise<CaptionOperationResponse> {
    try {
      const imagePath = path.resolve(request.image_path);
      const captionExt = request.caption_extension || DEFAULT_CAPTION_EXTENSION;
      const captionPath = imagePath.replace(path.extname(imagePath), captionExt);

      // Ensure parent directory exists
      const parentDir = path.dirname(captionPath);
      await fs.mkdir(parentDir, { recursive: true });

      // Write caption
      await fs.writeFile(captionPath, request.caption_text, 'utf-8');

      return {
        success: true,
        message: `Caption written to ${path.basename(captionPath)}`,
        files_modified: 1,
        errors: [],
      };
    } catch (error) {
      return {
        success: false,
        message: `Failed to write caption: ${error instanceof Error ? error.message : String(error)}`,
        files_modified: 0,
        errors: [error instanceof Error ? error.message : String(error)],
      };
    }
  }
}

// Export singleton instance
export const captionService = new CaptionService();
